packages/cli/src/services/FileDiscovery.ts

```diff
@@ -0,0 +1,155 @@
+import { existsSync, realpathSync } from 'node:fs';
+import { readdir } from 'node:fs/promises';
+import { join, normalize, relative, resolve } from 'node:path';
+import { DEFAULT_SRC_DIR } from '../config.ts';
+import { CompilationError } from '../types/errors.ts';
+
+/**
+ * Service responsible for discovering .compact files in the source directory.
+ * Recursively scans directories and filters for .compact file extensions.
+ *
+ * @class FileDiscovery
+ * @example
+ * ```typescript
+ * const discovery = new FileDiscovery('src');
+ * const files = await discovery.getCompactFiles('src/security');
+ * console.log(`Found ${files.length} .compact files`);
+ * ```
+ */
+export class FileDiscovery {
+  private srcDir: string;
+
+  /**
+   * Creates a new FileDiscovery instance.
+   *
+   * @param srcDir - Base source directory for relative path calculation (default: 'src')
+   */
+  constructor(srcDir: string = DEFAULT_SRC_DIR) {
+    this.srcDir = srcDir;
+  }
+
+  /**
+   * Validates and normalizes a file path to prevent command injection.
+   * Ensures the path exists, resolves symlinks, and is within allowed directories.
+   *
+   * @param filePath - The file path to validate
+   * @param allowedBaseDir - Base directory that the path must be within
+   * @returns Normalized absolute path
+   * @throws {CompilationError} If path is invalid or contains unsafe characters
+   * @example
+   * ```typescript
+   * const discovery = new FileDiscovery('src');
+   * const safePath = discovery.validateAndNormalizePath(
+   *   'src/MyToken.compact',
+   *   'src'
+   * );
+   * ```
+   */
+  validateAndNormalizePath(filePath: string, allowedBaseDir: string): string {
+    // Normalize the path to resolve '..' and '.' segments
+    const normalized = normalize(filePath);
+
+    // Check for shell metacharacters and embedded quotes
+    if (/[;&|`$(){}[\]<>'"\\]/.test(normalized)) {
+      throw new CompilationError(
+        `Invalid file path: contains unsafe characters: ${filePath}`,
+        filePath,
+      );
+    }
```

Comment:
This is unnecessary when using the execFile API (which is the default). No shell is spawned that can interpret these characters.
```suggestion
```